text/csv-indexation.md

```diff
@@ -0,0 +1,127 @@
+```
+
+*Example with an array of value inside a cell*
+```
+"id","label","price","colors","description"
+"1","t-shirt","4.99","red|green|blue","Hey, you will ""rock"" at summer time."
+```
+*should result to*
+```json=
+{
+  "hits": [
+    {
+      "id": 1,
+      "label": "t-shirt",
+      "price": 4.99,
+      "colors": ["red", "green", "blue"],
+      "description": "Hey, you will rock at summer time.",
+    }
+  ],
+  ...
+}
+```
+
+
+### Impact on documentation
+
+This feature should impact MeiliSearch users documentation by adding mention of csv capability inside Documents scope at [Add or replace documents](https://docs.meilisearch.com/reference/api/documents.html#add-or-replace-documents) and [Add or update documents](https://docs.meilisearch.com/reference/api/documents.html#add-or-update-documents). Documentation should also guide the user in the correct way to properly format and send the csv file. Adding a dedicated page for the purpose of formatting and sending a CSV file should be considered.
+
+### Impact on SDKs
+
+This feature should impact MeiliSearch SDK's by adding the possibility to send a CSV file onto the http protocol for the previous explicited endpoints.
+
+>Note that SDKs should add a specific ***Content-Type*** text/csv header on POST and PUT Documents endpoints.
+
+## Technical Specifications
+
+### Architecture
+
+### Technical Limitations
+
+- Is CSV payload size is equivalent to [JSON payload size](https://docs.meilisearch.com/reference/features/configuration.html#payload-limit-size)?
+>Does handling the process of streaming chunks of a huge file is conceivable for now?
+
+### Corner Cases
+
+- Handling an array of values in a cell seems to be far from being easily to handle at file format step. Therefore, it's not providing a good usage experience of MeiliSearch. One way to do it can be to transform csv rows to json objects internally and send it to Milli. We could use two http parameters to handle the logic related to multiple values separation. One for marking the columns to be separated by checking presence of a delimiter and another one to explicitly set this delimiter. This is what Algolia use in [algoliaCsvTools](https://github.com/algolia/algolia-csv-js) to facilitate csv file uploads from CLI. See ***parseArrays*** and ***arrayDelimiter*** parameters. ***We can use only one parameter by enforcing a delimiter.***
+- Should we privilege CSV ingestion at engine level (avoiding messaging transplant with http call leading to network overhead) or CSV ingestion from SDk's point of view for developper experience. Also, the last point allows building of an interface in the dashboard to send CSV file to the search engine. Should we considerate a first interation aims at feature adoption and payload size analytics? If we advocate for the discovery aspect of the feature, we should keep it as a simple and stupid (not handle the corner case of cell containing array definition)
```

Comment:
I'm not sure what you mean by "avoid messaging transplant"

Reply:
When you say ingest csv from the sdk point of view, you mean from the API?

Reply:
> We want to privilege a data format that is streamable and that does not require unmarshalling to disk: both csv and jsonl are supported now that allow that. The conversion to json is done at search time, when the documents are federalized from.the database

It seems that Algolia proposes a tool to convert complex CSV values (For example the cell "genres" contains values that need to be parsed as an array). To do that, they seem to convert CSV to JSON and push it to their main JSON indexing method. The main caveat of that is concerning the performance issue since we are aiming to propose a fast way of importing a bigger dataset. Does the CSV indexation method currently handle that type of cell from a CSV file?

> I'm not sure what you mean by "avoid messaging transplant"
I should have elaborate on my question more precisely. Please correct me if I'm wrong but I understand that Transplant is used to expose HTTP API Endpoints. By avoiding Transplant, I mean using the CLI on the engine side.

>When you say ingest csv from the sdk point of view, you mean from the API?
Exactly.

Reply:
>  Does the CSV indexation method currently handle that type of cell from a CSV file?
No, it only support ISO csv. We use this lib under the hood: https://docs.rs/csv/1.1.6/csv/
The best thing to do would be to convert the csv into jsonl and not json, so we keep the streaming capabilities and performance. Handling such csv is a real question for me though.

> I should have elaborate on my question more precisely. Please correct me if I'm wrong but I understand that Transplant is used to expose HTTP API Endpoints.

transplant does a lot more than that, I'll explain it to you on our call this afternoon if you want :)

> Exactly.

gotcha, we usually use sdk to refer to the sdk integrations :)